File: leetcode3/최원준/481. Magical String.py
#

'''
1. 아이디어 :
_next가 가리키는 값(1 또는 2)을 현재 숫자 num을 몇 번 붙일지로 사용.
처음에는 고정 start [1, 2]를 읽어서 1, 22를 만들고, 이후부터는 만들어진 s 자신을 읽는다.
num은 매번 1 → 2 → 1 → 2로 바꾸고, s 길이가 n이 되면 1의 개수를 센다.

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :


'''
class Solution:
    def magicalString(self, n: int) -> int:
        # 1 22 11 2 1 22 1 22 11 2 11 22
        # 1 22 11 2 1 22 1 22
        # 1 22 11 2 1 2
        # 1 22 11
        # 1 22
        # 1 2
        
        s = []
        start = [1, 2]

        _next = 0
        num = 1

        while len(s) < n:
            if _next < len(start):
                count = start[_next]
            else:
                count = s[_next]

            for _ in range(count):
                s.append(num)
                if len(s) == n:
                    break
                # print(s)
            
            num = 3 - num
            _next += 1
        return s.count(1)
        

File: leetcode3/이진희/481_Magical String.java
/*

1. 아이디어 : 특별한 문자열을 직접 구해 1의 개수를 계산한다
              1. 숫자는 1부터 시작한다
              2. s = "1221121221221121122……" 로 이미 앞부분은 정해져 있다
              3. 3번째 숫자 2부터 직전숫자가 1이면 2, 2면 1이라는 규칙을 가지고, 문자열을 만든다
              4. n이되면 1의 개수를 return 한다

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : 완전탐색, 구현

 */

class Solution {
    public int magicalString(int n) {
        // 최대 10만

        int idx = 3;
        int ans = 1;
        int pointer = 2;

        if(n == 1) return ans;

        StringBuilder sb = new StringBuilder();
        sb.append(122);
        int addNum = 0;

        while(true) {
            int num = sb.charAt(pointer)-'0';
            if(sb.charAt(sb.length()-1) == '2') addNum = 1;
            else addNum = 2;

            while(num>0){
                if(idx >= n) return ans;
                if(addNum == 1) ans++;
                sb.append(addNum);
                idx++;
                num--;
            }

            pointer++;
        }
    }
}